### ISSUE TITLE
React Native (JavaScriptCore) Incompatibility - "Conflicting definitions of %InertAsyncFunction%"

### ISSUE THREAD

**Issue by EtDu** [endojs/endo#659]
When importing and calling `lockdown()` in a React Native context, the following error appears:
<img width="531" alt="Screen Shot 2021-04-08 at 11 42 51 AM" src="https://user-images.githubusercontent.com/26778340/113969571-cb03a800-9867-11eb-97dd-bc1621fcc2ec.png">

entry `index.js` file
```js
import { AppRegistry } from 'react-native';
import App from './App';
import { name as appName } from './app.json';
import './lockdown.umd.js'

lockdown({ errorTaming: 'unsafe'})

AppRegistry.registerComponent(appName, () => App);
```

ELI5 React Native Setup:
```
yarn install -g expo-cli
expo init LockdownBug
cd LockdownBug
```

Add `lockdown.umd.js` in the project root
Then, inside `index.js`:
```js
import { registerRootComponent } from 'expo';

import App from './App';
import './lockdown.umd'
lockdown({errorTaming: 'unsafe'})

// registerRootComponent calls AppRegistry.registerComponent('main', () => App);
// It also ensures that whether you load the app in the Expo client or in a native build,
// the environment is set up appropriately
registerRootComponent(App);

```

Finally,
```
yarn web
```

This opens a debugging tab with the error present!

**Comment by kumavis** [comment]
on `react-native` the `%InertAsyncFunction%` intrinsic is added
- first here https://github.com/endojs/endo/blob/70cc86eb400655e922413b99c38818d7b2e79da0/packages/ses/src/lockdown-shim.js#L182
- second here (causes error) https://github.com/endojs/endo/blob/70cc86eb400655e922413b99c38818d7b2e79da0/packages/ses/src/lockdown-shim.js#L191

**Comment by kumavis** [comment]
this happens normally, but the error is only thrown when the prop descs dont deep equal each other.
On `react-native` (JavascriptCore)
- the value from `tameFunctionConstructors` is `AsyncFunction`
- the value from `getAnonymousIntrinsics` is `Function` <------- this differs from node

**Comment by kumavis** [comment]
we're seeing
```js
Object.getPrototypeOf(async function (){}) === Fuunction.prototype //=> true
```
on this stack:
`react-native`: `0.62.2`
`apple SimulatorKit simulator`: `12.4`
`simulated 8th gen iPad iOS`: `14.4`

we're only seeing this on react-native for ios, which uses the system's JavaScriptCore version
we're **not** seeing this on react-native android builds, which bundle their own version of JavaScriptCore
we're **not** seeing this on Safari

**Comment by kumavis** [comment]
heres the error in `react-native` on android via the android simulator
notice that it says "js-engine: v8"
![image](https://user-images.githubusercontent.com/1474978/114128524-1df46280-992f-11eb-93de-61b7ed14110a.png)

**Comment by kumavis** [comment]
verified that lockdown works on ios safari in the apple simulator

**Comment by kumavis** [comment]
increasingly seems like a weird `react-native`-specific thing and not a JavaScriptCore issue

**Comment by kumavis** [comment]
ah hah ❗ 
react-native is transforming `lockdown.umd.js`, replacing the call to `async function (){}` with a non-async function that uses `regenerator-runtime`. This transformation is not happening for `tameFunctionConstructors` because it creates its async function via eval https://github.com/endojs/endo/blob/70cc86eb400655e922413b99c38818d7b2e79da0/packages/ses/src/tame-function-constructors.js#L113-L117

here is an example of the transformation.

original:
```js
const testResult = Object.getPrototypeOf(async function AsyncFunctionInstance() {}) === Function.prototype
//=> normally false, but true after transformation
```

after transformation:
```js
var testResult = Object.getPrototypeOf(function AsyncFunctionInstance() {
  return regeneratorRuntime.async(function AsyncFunctionInstance$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
        case "end":
          return _context.stop();
      }
    }
  }, null, null, null, Promise);
}) === Function.prototype;
//=> true
```

**Comment by kumavis** [comment]
here is an browser sandbox that applies the transformation
https://snack.expo.io/rJjV44yyJ

**Comment by kumavis** [comment]
to configure react-native metro bundler to not transform SES

`babel.config.js`
```js
const path = require('path')
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
  ignore: [
    path.dirname(require.resolve('ses/package.json'))
  ]
};
```

this did **not** work
```js
ignore: [require.resolve('ses/lockdown')]
```

remember to `--reset-cache` so that transforms are properly applied

**Comment by kumavis** [comment]
closing as this is technically a react-native configuration error

**Comment by bentobox19** [comment]
Just updating the steps to replicate the solution from a blank react native application run with `expo`

### Summary

* Set app a blank application
* Add `lockdown.umd.js` to your app
* Find out entry point in `package.json`
* Add dependency and call `lockdown()`
* Modify `babel.config.js` to avoid transforming `lockdown.umd.js`
* Run with `yarn web`

### Steps

* Set up a blank application

```bash
expo init blah
cd blah
```

* Add `lockdown.umd.js` (from https://npmfs.com/download/ses/0.15.2/dist/lockdown.umd.js)

```bash
curl -O https://npmfs.com/download/ses/0.15.2/dist/lockdown.umd.js
```

* See that `package.json` has the entry

```json
"main": "index.js",
```

* `index.js`

```js
import { registerRootComponent } from 'expo';

import App from './App';
import './lockdown.umd'
lockdown({errorTaming: 'unsafe'})

// registerRootComponent calls AppRegistry.registerComponent('main', () => App);
// It also ensures that whether you load the app in the Expo client or in a native build,
// the environment is set up appropriately
registerRootComponent(App);
```

* You need to modify `babel.config.js`

```js
module.exports = function(api) {
  api.cache(true);
  return {
    presets: ['babel-preset-expo'],
    ignore: ['./lockdown.umd.js']
  };
};
```

* Finally  `yarn web`
